fix(aws-lambda): sanitize non-ASCII header values to prevent ByteString errors

yusukebe opened the pull request:
Fixes #4432

### The author should do the following, if applicable

- [x] Add tests
- [x] Run tests
- [ ] `bun run format:fix && bun run lint:fix` to format the code
- [ ] Add [TSDoc](https://tsdoc.org/)/[JSDoc](https://jsdoc.app/about-getting-started) to document the code

ajaishankar commented:
Hi @yusukebe I wanted to refer to the Cloudflare documentation

https://developers.cloudflare.com/rules/transform/managed-transforms/reference/#encoding-of-non-ascii-header-values

> Cloudflare always converts non-ASCII characters to UTF-8 (using hexadecimal character representation) in HTTP request and response header values.
>
> For example, the header value São Paulo would be encoded as S\u00c3\u00a3o Paulo.

What I _think_ this means is Cloudflare is supposed to send `São Paulo` as `'S%C3%A3o%20Paulo'` (hexadecimal).

When I raised the issue our ALB was receiving this string `"Srīperumbūdūr"` for `cf-ipcity` header from Chennai, India. 

Maybe it's a bug with Cloudflare as it did not encode this city?

As for the fix, can't Hono take an approach similar to Cloudflare?

- If there is a char > 255 simply `encodeURIComponent` the value
- Leave it to the client code to try `decodeURIComponent` if say it finds a % in the value?

Not sure if this is a good approach, but won't decoding as `latin1` as in this PR return a wrong string?

yusukebe commented:
Hi @ajaishankar 

I don't know whether it's a bug with Cloudflare. But how about the current code? I added the change https://github.com/honojs/hono/pull/4437/commits/c51c060398deb7a2a71f82875cad35790da71d29, simplified and used `encodeURIComponent`.

ajaishankar commented:
Thanks @yusukebe! This looks good 👍

yusukebe commented:
@ajaishankar Thank you for reviewing. Merging.
